2020 Resource Hints chapter

Zizzamia opened the pull request:
In this PR we introduce the first official draft as markdown for the **Resource Hints 2020** chapter #920.

Following PRs will focus on:
- polishing figures
- review all backlinks
- possible final editing

Thank you @rviscomi and @obto for all the guidance, and all reviewers and analysts (@notwillk, @giopunt, @jessnicolet, @pmeenan, @mgechev) for the incredible feedbacks and help. 🙏
Please, feel free to keep mentioning anything in terms of content or data we might want to polish, I am going to use the next few days to do several final clean-up anyway. 🌲 🚀 🌕

rviscomi commented on src/content/en/2020/resource-hints.md:
```diff
@@ -8,16 +8,336 @@ authors: [Zizzamia]
+The largest use is with script elements, which is unsurprising as the number of JS primary and third-party files continues to grow.
+
+{{ figure_markup(
+  caption="Priority hints on mobile have &quot;low&quot; importance",
```
```suggestion
  caption="Priority hints on mobile have \"low\" importance",
```

Zizzamia replied:
Nice, great catch! I just added the fix requested.

rviscomi reviewed (changes requested):
One quick suggestion to get the build working

I've staged the draft here to assist with the review: https://20201127t212713-dot-webalmanac.uk.r.appspot.com/en/2020/resource-hints

bazzadp commented:
Looks good to me @Zizzamia but difficult to be sure until we see the charts in place. I would normally advocate for smaller PRs and an iterative approach but, since the chapter isn't actually launch-able without the charts I think we should change this to a Draft PR, and then hold it open until the charts are added and the TODOs are finished out.

BTW, just a heads up, after we merge this, there will be a round of editing from someone outside the immediate team (@rviscomi or myself most likely unless we can get some more volunteer Editors) 😁 to try to bring a consistent style and voice across all the chapters. This can be substantial just to warn you as some Authors are surprised with the level of editing. You'll will be asked to review all suggested edits to make sure you are comfortable with them.

It is possible to launch the chapter with an unedited tag (see the 2020 Markup Chapter](https://almanac.httparchive.org/en/2020/markup) as an example, but that is the minimal viable product of the chapter so do really need the images to get to that:

![2020 Markup Chapter with Unedited label](https://user-images.githubusercontent.com/10931297/100518427-4dc57980-3189-11eb-9f90-f88790926988.png)

rviscomi commented on src/content/en/2020/resource-hints.md:
```diff
@@ -8,16 +8,336 @@ authors: [Zizzamia]
 reviewers: [notwillk, giopunt, jessnicolet, pmeenan, mgechev]
 analysts: [khempenius]
```
Can you verify that all of these reviewers/analysts have meaningfully contributed to the chapter and are roughly in sorted order so that the biggest contributors are listed first?

rviscomi commented on src/content/en/2020/resource-hints.md:
```diff
@@ -8,16 +8,336 @@ authors: [Zizzamia]
+
+More and more web pages are using the main Resource Hints, and in 2020 we are seeing the adoption remains incredibly consistent between desktop & mobile.
+
+// TODO @zizzamia Add Chart
```
I've published the chart to generate the embed URL and gotten the figure started but left out the image and caption/description.
```suggestion
{{ figure_markup(
  image="TODO",
  caption="TODO",
  description="TODO",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTYAbLxN40s6mNR1jo0XDe_V4siN8TAsx2mryMp5IQmlJ-9O9eJxYROz7Rw6ozyFP6hlIZHxxh95GqX/pubchart?oid=1550112064&format=interactive",
  sheets_gid="1805612941",
  sql_file="adoption.sql"
) }}
```

rviscomi commented on src/content/en/2020/resource-hints.md:
```diff
@@ -8,16 +8,336 @@ authors: [Zizzamia]
+
+Across the board developers are learning how to better use Resource Hints, and compared to [2019](https://almanac.httparchive.org/en/2019/resource-hints#resource-hints) we've seen an improved use of `preload`, `prefetch` and `preconnect`. For expensive operations like preload and preconnect the median usage on desktop decreased from 2 to 1. We have seen the opposite for loading future resources w ...
+
+// TODO @zizzamia Add chart 2
```
```suggestion
{{ figure_markup(
  image="TODO",
  caption="TODO",
  description="TODO",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTYAbLxN40s6mNR1jo0XDe_V4siN8TAsx2mryMp5IQmlJ-9O9eJxYROz7Rw6ozyFP6hlIZHxxh95GqX/pubchart?oid=320451644&format=interactive",
  sheets_gid="175042082",
  sql_file="hints_per_page.sql"
) }}
```

rviscomi commented on src/content/en/2020/resource-hints.md:
```diff
@@ -8,16 +8,336 @@ authors: [Zizzamia]
+  caption="The most preload hints per page 🤯",
+  content="20,931",
+  classes="big-number",
+  sheets_gid="1805612941",
```
This gid value should correspond to the one in the figure above: https://docs.google.com/spreadsheets/d/1lXjd8ogB7kYfG09eUdGYXUlrMjs4mq1Z7nNldQnvkVA/edit#gid=175042082

If these values are just placeholders could you update the other figures to use TODOs instead?
```suggestion
  sheets_gid="175042082",
```

rviscomi commented on src/content/en/2020/resource-hints.md:
```diff
@@ -8,16 +8,336 @@ authors: [Zizzamia]
+
+This is likely related to a large group of sites built as Single Page Apps that need the main bundle as soon as possible to start downloading the rest of their JS dependencies. Subsequent usage comes from font at 8%, style at 5%, image at 1%, and fetch at 1%.
+
+// TODO @zizzamia Add chart 3
```
```suggestion
{{ figure_markup(
  image="TODO",
  caption="TODO",
  description="TODO",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTYAbLxN40s6mNR1jo0XDe_V4siN8TAsx2mryMp5IQmlJ-9O9eJxYROz7Rw6ozyFP6hlIZHxxh95GqX/pubchart?oid=903180926&format=interactive",
  sheets_gid="1829901599",
  sql_file="as_attribute_by_year.sql"
) }}
```

rviscomi commented on src/content/en/2020/resource-hints.md:
```diff
@@ -8,16 +8,336 @@ authors: [Zizzamia]
+
+For both `preload` and `prefetch` we've had an increase in adoption when the page is controlled by a [Service Worker](https://developers.google.com/web/fundamentals/primers/service-workers). This is because of the potential to both improve the resource prioritization by preloading when the Service Worker is not active yet and intelligently prefetching future resources while letting the Service Wo ...
+
+// TODO @zizzamia Add chart 4
```
```suggestion
{{ figure_markup(
  image="TODO",
  caption="TODO",
  description="TODO",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTYAbLxN40s6mNR1jo0XDe_V4siN8TAsx2mryMp5IQmlJ-9O9eJxYROz7Rw6ozyFP6hlIZHxxh95GqX/pubchart?oid=252958553&format=interactive",
  sheets_gid="691299508",
  sql_file="adoption_service_workers.sql"
) }}
```

Zizzamia commented:
@bazzadp thank you for the initial push-back, with @rviscomi  we added the charts into this PRs. Also we were able to update the order of reviewers based on the amount of contributions and added all the correct `gid` and `sql` values into the figures.

If it's ok for the team I will follow up with more _polishing figures_, _backlinks review_ and _possible final editing_ in the next PRs.

And yeah, looking forward from all editors to review and re-shape the article in a way it's more aligned with the rest of the Almanac 2020. 🌲🌕🌲

bazzadp commented on src/content/en/2020/resource-hints.md:
````diff
@@ -5,19 +5,367 @@ chapter_number: 21
+<link rel="preload" href="mobile.css" as="style" media=”~~~”>
+```
+
+The most popular media value used with `preload` is a combination of `only screen and (max-width: 768px)` and `screen and (min-width: 767px)`. Seeing over 2,100 different combinations of media queries in the 2020 dataset encourages us to consider how wide the variance is between concept and implementation of responsive design from site to site.
````
I'm not sure the most popular values are worth commenting on unless significantly more than the others? I was intrigued enough by this sentence [to look at the data](https://docs.google.com/spreadsheets/d/1lXjd8ogB7kYfG09eUdGYXUlrMjs4mq1Z7nNldQnvkVA/edit#gid=2016228012) and doesn't seem like these two values have any real significance:

![preload media values](https://user-images.githubusercontent.com/10931297/100557228-a2f3af00-329f-11eb-95d8-239138274bf2.png)

```suggestion
Seeing over 2,100 different combinations of media queries in the 2020 dataset encourages us to consider how wide the variance is between concept and implementation of responsive design from site to site.
```

Zizzamia replied:
True, at the same time I find quite interesting the whole ecosystem is trying to conform on one particular group of number 768 and 767 to split desktop vs mobile.

I wonder how much the [breakpoints in w3schools](https://www.w3schools.com/css/css_rwd_mediaqueries.asp) contributed on this number.

What you think? Should we share something about this convergence or not?

bazzadp replied:
I'd say it's more likely to be due to the popularity of Bootstrap than w3schools!

But yes I think you can say something about this. Maybe the second sentence rather than the first will de-emphasise it. Something like:

```suggestion
Seeing over 2,100 different combinations of media queries in the 2020 dataset encourages us to consider how wide the variance is between concept and implementation of responsive design from site to site. The ever popular `767px/768px` breakpoints (as popularised by Bootstrap amongst others) can be seen in the data.
```

bazzadp commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+  sql_file="lighthouse_preload.sql"
+) }}
+
+Running Lighthouse’s "[Preload key requests](https://web.dev/uses-rel-preload/)" audit resulted in 84.6% of pages passing the test, which is an astonishing result. If you are looking to use `preload` for the first time, remember, fonts and critical scripts are a good place to start.
```
I think it's important to understand what this audit means, as it's a bit less astonishing then.

It does NOT mean 84.6% of pages are using preload successfully. It means either there are, or there are no third level requests.

Crucially [this audit only looks at 1st party requests](https://github.com/GoogleChrome/lighthouse/blob/5c5de46ffce484f969ea5df8d71bce26dbc27ed6/lighthouse-core/audits/uses-rel-preload.js#L100) so things like Google Fonts aren't included.

Given that we only look at home pages are they less likely to have bundles apps with dependencies?

Think you should consider rephrasing this section.

bazzadp commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+
+### Native Lazy Loading
+
+Now let’s celebrate the first year of the [Native Lazy Loading](https://addyosmani.com/blog/lazy-loading/) API, which at the time of publishing already has over [72%](https://caniuse.com/loading-lazy-attr) browser support. This new API can be used to defer the load of below-the-fold iframes and images on the page until the user scrolls near them. This can reduce data usage, memory usage, and help ...
```
Worth mentioning that Safari, as the last major browser to support this, has implemented this but not turned it on yet at which point support (for images at least) will jump into the high 90s?

Zizzamia replied:
Do you have a link we could use to mention this information?

bazzadp replied:
Can't find an official link (the [webkit bug](https://bugs.webkit.org/show_bug.cgi?id=196698) is marked as resolved

Let's just use caniuse: https://caniuse.com/loading-lazy-attr

Zizzamia replied:
Sounds good!

bazzadp commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+
+Now let’s celebrate the first year of the [Native Lazy Loading](https://addyosmani.com/blog/lazy-loading/) API, which at the time of publishing already has over [72%](https://caniuse.com/loading-lazy-attr) browser support. This new API can be used to defer the load of below-the-fold iframes and images on the page until the user scrolls near them. This can reduce data usage, memory usage, and help ...
+
+{{ figure_markup(
+  caption="Percentage of pages using native lazy loading set to “lazy”",
+  content="3.87%",
+  classes="big-number",
+  sheets_gid="2039808014",
+  sql_file="native_lazy_loading_attrs.sql"
+) }}
```
Is this images or images and iframes? Think it would be good to clarify (applies to this whole section). And if just images should we have a sentence or two on iframes? We use it on our site :-)

bazzadp commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+
+Be mindful to run the audit on both desktop and mobile as images may move off screen when the viewport changes.
+
+## Machine Learning
```
What do you think about changing this title to predictive prefetching? As Machine Learning is just one way or predicting this.

```suggestion
## Predictive prefetching
```

bazzadp commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+  sql_file="h2_preload_nopush.sql"
+) }}
+
+In the 2020 dataset we have seen 1% of mobile pages using HTTP/2 Push, and of those 75% of preload header links use the nopush option in the firstHtml request. This means that even though a website is using preload, the majority of requests prefer to disable the push option.
```
I think it's important to explain the relationship between preload and push.

```suggestion
HTTP/2 Push is often initiated through the `preload` link header. In the 2020 dataset we have seen 1% of mobile pages using HTTP/2 Push, and of those 75% of preload header links use the `nopush` option in the page request. This means that even though a website is using the `preload` resource hint, the majority prefer to use just this and disable HTTP/2 pushing of that resource.
```

Zizzamia replied:
Niiice, yeah that's much better! Thank you

bazzadp commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+
+In the 2020 dataset we have seen 1% of mobile pages using HTTP/2 Push, and of those 75% of preload header links use the nopush option in the firstHtml request. This means that even though a website is using preload, the majority of requests prefer to disable the push option.
+
+It's important to mention that HTTP/2 Push can also damage performance if not used correctly.
```
```suggestion
It's important to mention that HTTP/2 Push can also damage performance if not used correctly which probably explains why is often disabled.
```

bazzadp commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+
+## Future
+
+Let's dive into a couple of experimental hints. Very close to release we have Priority Hints, at the moment actively experimented with the web community. Lastly we have the 103 Early Hints in HTTP/2, which is still in early inception and there are a few players like Chrome and Fastly collaborating for upcoming test trials.
```
```suggestion
Let's dive into a couple of experimental hints. Very close to release we have Priority Hints, at the moment actively experimented with the web community. Lastly we have the 103 Early Hints in HTTP/2, which is still in early inception and there are a few players like [Chrome and Fastly collaborating for upcoming test trials]](https://www.fastly.com/blog/beyond-server-push-experimenting-with-the-103-early-hints-status-code).
```

bazzadp reviewed (changes requested):
Very well done @Zizzamia ! Great chapter and I enjoyed reading that.

I do have some comments for you to consider though.

bazzadp commented:
Oh and one more comment, there do seem to be an awful lot of the big callout figures in this chapter. Very much like resource hints, over use can diminish the impact of all of them 😁 Could we replace any with charts or tables?

Zizzamia commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
 analysts: [khempenius]
 translators: []
-#Zizzamia_bio: TODO
+zizzamia_bio: Staff Software Engineer, Retail Growth at Coinbase
```
Oh I wasn't aware was actually an option 😁

rviscomi replied:
It is! And by the way most authors usually write more about themselves than just title/employer. Try to answer the question: "Who are you and why are you qualified to write this chapter?"

bazzadp replied:
One thing to note is all the meta data is HTML-only. Really should add Markdown support as see some authors trying to use that, but that's for another days.

Zizzamia replied:
LOL, yeah I kind of put just my job as placeholder. I will give it a second pass in the next PR. thx @rviscomi

notwillk commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
 description: Resource Hints chapter of the 2020 Web Almanac covering usage of dns-prefetch, preconnect, preload, prefetch, priority hints, and native lazy loading.
 authors: [Zizzamia]
-reviewers: [notwillk, giopunt, jessnicolet, pmeenan, mgechev]
+reviewers: [jessnicolet, pmeenan, giopunt, mgechev, notwillk]
```
It looks like I'm in there (name is wrong).  Is up dating my name all I need to do there?

bazzadp replied:
Ah yes I saw this:

![Resource Hints contributors](https://user-images.githubusercontent.com/10931297/100592497-153fb000-32ef-11eb-9793-ead32b6f76c0.png)

And presumed you were in there. If you could add your full name (and any other details you want to show in the Contributors page) that would be great.

Zizzamia commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+
+Over the past decade [Resource Hints](https://www.w3.org/TR/resource-hints/) have become essential primitives that allow developers to improve page performance and therefore the user experience. 
+
+Preloading resources and having browsers apply some intelligent prioritization is something that was actually started way back in 2009 by IE8 with something called the [preloader](https://speedcurve.com/blog/load-scripts-async/). In addition to the HTML parser, IE8 had a lightweight look-ahead preloader that scanned for tags that could initiate network requests (script, link, images).
```
Right, I was referring to `<script>`, `<link>` and `<img>`. Great catch!

bazzadp replied:
```suggestion
Preloading resources and having browsers apply some intelligent prioritization is something that was actually started way back in 2009 by IE8 with something called the [preloader](https://speedcurve.com/blog/load-scripts-async/). In addition to the HTML parser, IE8 had a lightweight look-ahead preloader that scanned for tags that could initiate network requests (`<script>`, `<link>`, and `<img>`).
```

Zizzamia commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+
+### dns-prefetch
+
+[dns-prefetch](https://web.dev/preconnect-and-dns-prefetch/) helps resolve the IP address for a given domain ahead of time. As the [oldest](https://caniuse.com/link-rel-dns-prefetch) resource hint available, it uses minimal CPU and network resources, and helps the browser to avoid experiencing the "worst-case" delay for DNS resolution, which can be [over 1 second](https://www.chromium.org/develop ...
+
+Be mindful when using `dns-prefetch` as even if they are lightweight to do it's easy to exhaust browser limits for the number of concurrent in-flight DNS requests allowed (Chrome still has a [limit of 6](https://source.chromium.org/chromium/chromium/src/+/master:net/dns/host_resolver_manager.cc;l=353)).
```
It might be more correct say "compared to `preconnect`" but we can keep it generic if you prefer. What you think?

bazzadp replied:
Yup that sounds good. Updated suggestion ready for you to commit if happy.

Zizzamia commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+
+Be mindful when using `dns-prefetch` as even if they are lightweight to do it's easy to exhaust browser limits for the number of concurrent in-flight DNS requests allowed (Chrome still has a [limit of 6](https://source.chromium.org/chromium/chromium/src/+/master:net/dns/host_resolver_manager.cc;l=353)).
+
+### preconnect
```
Nice!

Zizzamia commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+
+Be mindful, depending on the [quality](https://github.com/andydavies/http2-prioritization-issues#current-status) of HTTP/2 prioritization of the CDN used, `prefetch` prioritization could either improve performance or make it slower, by over prioritizing `prefetch` requests and taking away important bandwidth for the initial load. Make sure to double check the CDN you are using and adapt to take i ...
+
+### Native Lazy loading
+
+The [native lazy loading](https://web.dev/browser-level-image-lazy-loading/) hint is a native browser API for deferring the load of offscreen images and iframes. By using it, assets that are not needed during the initial page load will not initiate a network request, this will reduce data consumption and improve page performance.
```
They sounds both correct to me, we might at this point just follow what the spec prefers https://html.spec.whatwg.org/multipage/urls-and-fetching.html#lazy-loading-attributes which is without hyphen.

bazzadp replied:
Let's stick with lazy loading without the hypen then! Also means less to change form this year and last year's chapter :-)

Zizzamia commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+
+The [native lazy loading](https://web.dev/browser-level-image-lazy-loading/) hint is a native browser API for deferring the load of offscreen images and iframes. By using it, assets that are not needed during the initial page load will not initiate a network request, this will reduce data consumption and improve page performance.
+
+Be mindful Chromium's implementation of lazy-loading thresholds logic historically has been quite [conservative](https://web.dev/browser-level-image-lazy-loading/#distance-from-viewport-thresholds), keeping the offscreen limit to 3000px. During the last year the limit has been actively tested and improved on to better align developer expectations, and ultimately moving the thresholds to 1250px.
```
Right right, def we should mention that!

Zizzamia commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+
+### Hints adoption
+
+More and more web pages are using the main Resource Hints, and in 2020 we are seeing the adoption remains incredibly consistent between desktop & mobile.
```
I went back and forward on this sentence, at first I was leaning towards mentioning "mobile" got a slight larger percentage in adoption. Which to be fair I tend to user some resource hints more in "mobile", to gain some extra boost. BUT they made me notice the dataset between "mobile" and "desktop" is not similar enough to actually tell us a real difference between the two.

I will say let's drop "incredible" and just mention the adoption is similar between the two.

bazzadp replied:
Sounds good to me.

```suggestion
More and more web pages are using the main Resource Hints, and in 2020 we are seeing the adoption remains consistent between desktop & mobile.
```

Zizzamia commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+  sql_file="adoption.sql"
+) }}
+
+This is a great representation of how year over year we continue to focus on improving not only for desktop but also for mobile experiences. One take we will probably analyze even further next year is how HighEnd vs LowEnd devices will diverge in hints usage.
```
Yeah, I feel the same. That's why I think next year we should focus a bit less of desktop vs mobile but more on HighEnd vs LowEnd. But for 2020 it is still a stretch for developer to dedicate time on mobile, imagine to dedicate time to LowEnd as well. LOL

Happy to rephrase this to just share the facts and remove any mobile vs desktop observation.

Zizzamia commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+
+The relative popularity of `dns-prefetch` is unsurprising as it first appeared in 2009, and has the widest support out of all major Resource Hints. It had a 4% increase in Desktop adoption, compared to [2019](https://almanac.httparchive.org/en/2019/resource-hints#resource-hints). We saw a similar increase for `preconnect` as well. One key reason this was the largest growth between all hints, is t ...
+
+`preload` usage has had a slower growth with only a 2% increase from 2019. This could be in part because it requires a bit more attention. While you only need the domain to use `dns-prefetch` and `preconnect`, you must specify the resource to use `preload`. While you can safely assume `dns-prefetch` and `preconnect` will improve performance, if `preload` is used on the wrong resource, it can actu ...
```
They do, only if used correctly. Let me see if I can find some articles even from the past with good benchmarks.

bazzadp replied:
Yeah my point is the "if used correctly". I think the current phrasing goes against what was said above.

What about this phrasing?:

```suggestion
`preload` usage has had a slower growth with only a 2% increase from 2019. This could be in part because it requires a bit more attention. While you only need the domain to use `dns-prefetch` and `preconnect`, you must specify the resource to use `preload`. While `dns-prefetch` and `preconnect` are reasonably low risk–though still can be abused– `preload` has a much greater potential to actually damage performance if used incorrectly.
```

Zizzamia commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+  sql_file="hints_per_page.sql"
+) }}
+
+Resource hints are most effective when they're used selectively ("when everything is important, nothing is"). Having a more clear definition of what resources help improve Largest Contentful Paint moved the focus away from using `preconnect` and more towards `prefetch` by shifting some of the resource prioritization and freeing up bandwidth for what most helps the user at first.
```
Yeah I can tell my point is not clear at all here. What I am trying to say it's less about the kind of resource but more about the timing when we make the prioritization. A developer that use `preload` for everything it could end up to damage LCP because we are over-prioritizing resources that are not important for LCP where instead we should move some of those late resources to a later stage by using `prefetch`.

I think mentioning `prefetch` here creates confusion, we can def drop it. But I still believe we should teach developers to be aware of the network timeline on when those resources are need it and how they impact LCP.

bazzadp replied:
Oh OK. The specific use of LCP confused me. But you're using that to mean "initial rendering" I think?

What about something like this then?

```suggestion
Resource hints are most effective when they're used selectively ("when everything is important, nothing is"). Having a more clear definition of what resources help improve critical rendering, versus future navigation optimizations, can move the focus away from using `preconnect` and more towards `prefetch` by shifting some of the resource prioritization and freeing up bandwidth for what most helps the user at first.
```

Zizzamia commented on src/content/en/2020/resource-hints.md:
```diff
@@ -5,19 +5,367 @@ chapter_number: 21
+
+### The `crossorigin` attribute
+
+With `preload` and `preconnect` resources that have CORS enabled, such as fonts or images, it's important to include the `crossorigin` attribute, in order for the resource to be properly used. If the crossorigin attribute is absent, the request will follow the single-origin policy thereby making the use of preload useless.
```
That's a great question, I will say generally no!
Based on [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) the images edge cases is more for:
- Images/video frames drawn to a canvas using drawImage().
- CSS Shapes from images.

bazzadp replied:
I think as those are edge cases I wouldn't explicitly mention it and leave it as:

```
With `preload` and `preconnect` resources that have CORS enabled, such as fonts, it's important to include the `crossorigin` attribute, in order for the resource to be properly used. If the `crossorigin` attribute is absent, the request will follow the single-origin policy thereby making the use of preload useless.
```

You do say "such as" anyway so it's not like we're saying this is a definitive list.

Zizzamia commented:
@bazzadp thank you for reviewing, and yeah I would love to replace a few figures with tables and even add a bit more code examples. Would you like I made all future changes in this PR on in the next one (including a few comments you left as well)?

rviscomi commented:
I'd like to see us merge and iterate to unblock the editing step. If there's anything else that needs to be changed please leave TODO's in the markdown and resolve as suggestions in the editing PR itself, or a follow-up PR.

bazzadp commented:
> I'd like to see us merge and iterate to unblock the editing step. If there's anything else that needs to be changed please leave TODO's in the markdown and resolve as suggestions in the editing PR itself, or a follow-up PR.

Agreed. @Zizzamia could you have one last pass through all the open comments and either resolve them or add a Jinja2 comment (using `{# and #}`), which won't show up in the final HTML. Something like this:

```
{# TODO - clarify img versus img in this section - Ref https://github.com/HTTPArchive/almanac.httparchive.org/pull/1587#discussion_r532292106 #}
```

We can then edit this chapter and come back to these few remaining points later.

Zizzamia commented:
Todos in place, I'm going to work on them in the next PR 🌲🚀🌕
cc @rviscomi @bazzadp

rviscomi commented:
Ok great!
